File: cpp/tests/streaming/test_allgather.cpp
/**
 * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
 * SPDX-License-Identifier: Apache-2.0
 */

#include <cstring>
#include <numeric>

#include <cuda_runtime_api.h>
#include <gmock/gmock.h>
#include <gtest/gtest.h>

#include <rapidsmpf/communicator/single.hpp>
#include <rapidsmpf/cuda_stream.hpp>
#include <rapidsmpf/integrations/cudf/partition.hpp>
#include <rapidsmpf/memory/buffer.hpp>
#include <rapidsmpf/memory/packed_data.hpp>
#include <rapidsmpf/streaming/chunks/packed_data.hpp>
#include <rapidsmpf/streaming/coll/allgather.hpp>
#include <rapidsmpf/streaming/core/channel.hpp>
#include <rapidsmpf/streaming/core/context.hpp>
#include <rapidsmpf/streaming/core/leaf_node.hpp>
#include <rapidsmpf/streaming/core/node.hpp>

#include "base_streaming_fixture.hpp"

#include <coro/latch.hpp>

using namespace rapidsmpf;

class StreamingAllGather
    : public BaseStreamingFixture,
      public ::testing::WithParamInterface<std::tuple<int, MemoryType>> {
  public:
    void SetUp() override {
        BaseStreamingFixture::SetUpWithThreads(std::get<0>(GetParam()));
        GlobalEnvironment->barrier();
    }

    void TearDown() override {
        GlobalEnvironment->barrier();
        BaseStreamingFixture::TearDown();
    }
};

INSTANTIATE_TEST_SUITE_P(
    StreamingAllGather,
    StreamingAllGather,
    ::testing::Combine(
        ::testing::Values(1, 4), ::testing::Values(MemoryType::HOST, MemoryType::DEVICE)
    ),
    [](testing::TestParamInfo<StreamingAllGather::ParamType> const& info) {
        return "nthreads_" + std::to_string(std::get<0>(info.param)) + "_mem_type_"
               + (std::get<1>(info.param) == MemoryType::HOST ? "HOST" : "DEVICE");
    }
);

TEST_P(StreamingAllGather, basic) {
    auto mem_type = std::get<1>(GetParam());
    auto allgather = streaming::AllGather(ctx, OpID{0});

    int size = ctx->comm()->nranks();
    int rank = ctx->comm()->rank();

    constexpr int n_inserts{100};

    coro::latch latch(n_inserts);
    auto insert = [&](int insertion_id, std::uint64_t sequence) -> coro::task<void> {
        std::vector<int> data(size);
        for (int i = 0; i < size; i++) {
            data[i] = (n_inserts * rank + insertion_id) * size + i;
        }

        auto br = ctx->br();
        auto buf = br->allocate(
            br->stream_pool().get_stream(),
            br->reserve_or_fail(data.size() * sizeof(int), mem_type)
        );
        buf->write_access([&](std::byte* buf_data, rmm::cuda_stream_view stream) {
            RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
                buf_data,
                data.data(),
                data.size() * sizeof(int),
                cudaMemcpyDefault,
                stream
            ));
        });
        auto meta = std::make_unique<std::vector<std::uint8_t>>(sizeof(int));
        std::memcpy(meta->data(), &size, sizeof(int));
        allgather.insert(sequence, PackedData{std::move(meta), std::move(buf)});
        latch.count_down();
        co_return;
    };
    auto insert_finished = [&]() -> coro::task<void> {
        co_await latch;
        allgather.insert_finished();
    };
    std::vector<int> result(size * size * n_inserts);

    auto extract = [&]() -> coro::task<void> {
        auto data = co_await allgather.extract_all(streaming::AllGather::Ordered::YES);
        std::size_t offset{0};
        for (auto& pd : data) {
            RAPIDSMPF_EXPECTS(
                pd.metadata->size() == sizeof(int), "Invalid metadata buffer size"
            );
            int msize;
            std::memcpy(&msize, pd.metadata->data(), sizeof(int));
            RAPIDSMPF_EXPECTS(msize == size, "Corrupted metadata value");
            RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
                result.data() + offset,
                pd.data->data(),
                pd.data->size,
                cudaMemcpyDefault,
                pd.data->stream()
            ));
            offset += msize;
            pd.data->stream().synchronize();
        }
    };

    std::vector<coro::task<void>> pipeline{};
    // Insertions can all run concurrently.
    for (int i = 0; i < n_inserts; i++) {
        pipeline.push_back(ctx->executor()->schedule(insert(i, i)));
    }
    pipeline.push_back(ctx->executor()->schedule(insert_finished()));
    pipeline.push_back(ctx->executor()->schedule(extract()));
    streaming::run_streaming_pipeline(std::move(pipeline));
    std::vector<int> expected(size * size * n_inserts);
    std::iota(expected.begin(), expected.end(), 0);
    EXPECT_EQ(expected, result);
}

TEST_P(StreamingAllGather, streaming_node) {
    auto mem_type = std::get<1>(GetParam());

    auto ch_in = ctx->create_channel();
    auto ch_out = ctx->create_channel();

    int size = ctx->comm()->nranks();
    int rank = ctx->comm()->rank();

    constexpr int n_inserts{100};
    std::vector<streaming::Message> input_messages;
    for (int insertion_id = 0; insertion_id < n_inserts; insertion_id++) {
        std::vector<int> data(size);
        for (int i = 0; i < size; i++) {
            data[i] = (n_inserts * rank + insertion_id) * size + i;
        }

        auto br = ctx->br();
        auto buf = br->allocate(
            br->stream_pool().get_stream(),
            br->reserve_or_fail(data.size() * sizeof(int), mem_type)
        );
        buf->write_access([&](std::byte* buf_data, rmm::cuda_stream_view stream) {
            RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
                buf_data,
                data.data(),
                data.size() * sizeof(int),
                cudaMemcpyDefault,
                stream
            ));
        });
        auto meta = std::make_unique<std::vector<std::uint8_t>>(sizeof(int));
        std::memcpy(meta->data(), &size, sizeof(int));
        input_messages.emplace_back(
            streaming::to_message(
                insertion_id,
                std::make_unique<PackedData>(std::move(meta), std::move(buf))
            )
        );
    }
    std::vector<streaming::Message> output_messages;
    std::vector<coro::task<void>> pipeline{};
    pipeline.push_back(ctx->executor()->schedule(
        streaming::node::push_to_channel(ctx, ch_in, std::move(input_messages))
    ));
    pipeline.push_back(
        ctx->executor()->schedule(streaming::node::allgather(ctx, ch_in, ch_out, OpID{0}))
    );
    pipeline.push_back(ctx->executor()->schedule(
        streaming::node::pull_from_channel(ctx, ch_out, output_messages)
    ));
    streaming::run_streaming_pipeline(std::move(pipeline));
    std::vector<int> actual(size * size * n_inserts);
    std::size_t offset{0};
    for (auto& msg : output_messages) {
        auto pd = msg.release<PackedData>();
        RAPIDSMPF_EXPECTS(
            pd.metadata->size() == sizeof(int), "Invalid metadata buffer size"
        );
        int msize;
        std::memcpy(&msize, pd.metadata->data(), sizeof(int));
        RAPIDSMPF_EXPECTS(msize == size, "Corrupted metadata value");
        RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
            actual.data() + offset,
            pd.data->data(),
            pd.data->size,
            cudaMemcpyDefault,
            pd.stream()
        ));
        offset += msize;
        pd.stream().synchronize();
    }
    std::vector<int> expected(size * size * n_inserts);
    std::iota(expected.begin(), expected.end(), 0);
    EXPECT_EQ(expected, actual);
}

File: python/rapidsmpf/rapidsmpf/tests/streaming/test_allgather.py
# SPDX-FileCopyrightText: Copyright (c) 2025-2026, NVIDIA CORPORATION & AFFILIATES.
# SPDX-License-Identifier: Apache-2.0

from __future__ import annotations

from contextlib import nullcontext
from typing import TYPE_CHECKING

import pytest

from rapidsmpf.streaming.chunks.packed_data import PackedDataChunk
from rapidsmpf.streaming.coll.allgather import AllGather, allgather
from rapidsmpf.streaming.core.actor import define_actor, run_actor_network
from rapidsmpf.streaming.core.leaf_actor import pull_from_channel, push_to_channel
from rapidsmpf.streaming.core.message import Message
from rapidsmpf.testing import generate_packed_data, validate_packed_data

if TYPE_CHECKING:
    from collections.abc import Awaitable

    from rapidsmpf.communicator.communicator import Communicator
    from rapidsmpf.streaming.core.actor import CppActor
    from rapidsmpf.streaming.core.channel import Channel
    from rapidsmpf.streaming.core.context import Context


def _make_chunk(context: Context, num_rows: int, offset: int) -> PackedDataChunk:
    stream = context.br().stream_pool.get_stream()
    return PackedDataChunk.from_packed_data(
        generate_packed_data(num_rows, offset, stream, context.br()),
        br=context.br(),
    )


def _validate_message(
    context: Context, msg: Message[PackedDataChunk], num_rows: int, offset: int
) -> None:
    packed = PackedDataChunk.from_message(msg, br=context.br()).to_packed_data()
    validate_packed_data(packed, num_rows, offset)


def test_allgather_actor(context: Context, comm: Communicator) -> None:
    if comm.nranks != 1:
        pytest.skip("Only support single-rank runs")

    num_rows = 1000
    op_id = 0
    inputs = [_make_chunk(context, num_rows, i * num_rows) for i in range(3)]
    actors = []

    ch1: Channel[PackedDataChunk] = context.create_channel()
    actors.append(
        push_to_channel(
            context, ch1, [Message(i, chunk) for i, chunk in enumerate(inputs)]
        )
    )

    ch2: Channel[PackedDataChunk] = context.create_channel()
    actors.append(allgather(context, comm, ch1, ch2, op_id, ordered=True))

    actor, deferred = pull_from_channel(context, ch2)
    actors.append(actor)
    run_actor_network(context, actors=actors)

    results = deferred.release()
    assert len(results) == len(inputs)
    for i, msg in enumerate(results):
        assert msg.sequence_number == i
        _validate_message(context, msg, num_rows, i * num_rows)


@define_actor()
async def generate_inputs(
    context: Context, ch: Channel[PackedDataChunk], num_rows: int, num_chunks: int
) -> None:
    for i in range(num_chunks):
        await ch.send(context, Message(i, _make_chunk(context, num_rows, i * num_rows)))
    await ch.drain(context)


@define_actor()
async def allgather_and_forward(
    context: Context,
    comm: Communicator,
    ch_in: Channel[PackedDataChunk],
    ch_out: Channel[PackedDataChunk],
    op_id: int,
    use_context_manager: bool,  # noqa: FBT001
) -> None:
    gather = AllGather(context, comm, op_id)
    cm = gather if use_context_manager else nullcontext(gather)
    with cm as ag:
        while (msg := await ch_in.recv(context)) is not None:
            chunk = PackedDataChunk.from_message(msg, br=context.br()).to_packed_data()
            ag.insert(msg.sequence_number, chunk)
    if not use_context_manager:
        gather.insert_finished()
    gathered = await gather.extract_all(context, ordered=True)
    for sequence, packed in enumerate(gathered):
        await ch_out.send(
            context,
            Message(
                sequence,
                PackedDataChunk.from_packed_data(packed, br=context.br()),
            ),
        )
    await ch_out.drain(context)


@pytest.mark.parametrize(
    "use_context_manager", [True, False], ids=["context", "non-context"]
)
def test_allgather_object_interface(
    context: Context,
    comm: Communicator,
    use_context_manager: bool,  # noqa: FBT001
) -> None:
    if comm.nranks != 1:
        pytest.skip("Only support single-rank runs")

    ch_in: Channel[PackedDataChunk] = context.create_channel()
    ch_out: Channel[PackedDataChunk] = context.create_channel()
    actors: list[CppActor | Awaitable[None]] = []
    num_rows = 100
    num_chunks = 10
    op_id = 0
    actors.append(generate_inputs(context, ch_in, num_rows, num_chunks))
    actors.append(
        allgather_and_forward(context, comm, ch_in, ch_out, op_id, use_context_manager)
    )

    actor, deferred = pull_from_channel(context, ch_out)
    actors.append(actor)

    run_actor_network(context, actors=actors)
    results = deferred.release()
    assert len(results) == num_chunks
    for i, msg in enumerate(results):
        assert msg.sequence_number == i
        _validate_message(context, msg, num_rows, i * num_rows)
